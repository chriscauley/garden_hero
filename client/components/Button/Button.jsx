import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import { buildClassName } from 'utils/util';
import { buttonProps } from 'utils/propTypeEnum';

class Button extends Component {
  static propTypes = {
    className: PropTypes.string,
    children: PropTypes.node.isRequired,
    disabled: PropTypes.bool,
    fullWidth: PropTypes.bool,
    modifier: PropTypes.oneOf(buttonProps.modifiers),
    reference: PropTypes.func,
    size: PropTypes.oneOf(buttonProps.sizes),
    type: PropTypes.oneOf(buttonProps.types)
  };

  static defaultProps = {
    disabled: false,
    fullWidth: false,
    type: 'button'
  };

  render() {
    const {
      children,
      className,
      disabled,
      fullWidth,
      modifier,
      reference,
      size,
      type,
      ...props
    } = this.props;
    const mainClass = 'bcp-btn';
    const modifierClass = buildClassName(
      buttonProps.modifiers,
      modifier,
      mainClass
    );
    const disabledClass = disabled ? 'bcp-btn--disabled' : false;
    const sizeClass = buildClassName(buttonProps.sizes, size, mainClass);

    return (
      <button
        className={classNames(
          mainClass,
          modifierClass,
          disabledClass,
          sizeClass,
          { 'bcp-btn--full-width': fullWidth },
          className
        )}
        ref={reference}
        type={type}
        {...props}
      >
        {children}
      </button>
    );
  }
}

export default Button;
