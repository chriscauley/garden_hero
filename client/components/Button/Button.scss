.bcp-btn {
  @include body-copy;
  display: block;
  width: 100%;
  background-color: transparent;
  border: 0;
  padding: 0;
  cursor: pointer;
  @include responsive($widthMedium) {
    display: inline-block;
    width: auto;
  } // Adds Spacing between sibling buttons only on mobile
  &~& {
    margin-top: em(15px);
    @include responsive($widthMedium) {
      margin-top: 0;
    }
  }
  &--anchor,
  &--anchor-block {
    color: $blue;
  }
  &--anchor {
    display: inline-block;
    width: auto;
    text-align: left;
  }
  &--anchor-block {
    display: block;
    width: 100%;
    padding: 1.158em;
  }
  &--primary,
  &--secondary {
    min-width: 9em;
    padding: 0.684em 1.579em 0.632em;
    border-radius: 5em;
    &.bcp-btn--s {
      padding: 0.625em 1.875em;
    }
  }
  &--primary {
    background-color: $blue;
    color: $white;
    &:hover,
    &:focus,
    &:active {
      background-color: $blueBright;
    }
    &[disabled] {
      background-color: $gray;
    }
  }
  &--secondary {
    border: 1px solid $blue;
    color: $blue;
    &:hover,
    &:focus,
    &:active {
      border-color: $blueBright;
      color: $blueBright;
    }
  }
  &--disabled {
    background-color: $gray;
    cursor: default;
    &:hover,
    &:focus,
    &:active {
      background-color: $gray;
    }
    &.bcp-btn--secondary {
      background-color: transparent;
      border-color: $grayDark;
      color: $grayDark;
    }
  }
  &--full-width {
    width: 100%;
  }
  &--xl {
    @include heading-xl;
  }
  &--l {
    @include heading-l;
  }
  &--s {
    @include heading-s;
  }
}