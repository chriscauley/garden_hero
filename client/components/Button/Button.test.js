import React from 'react';
import renderer from 'react-test-renderer';

import Button from './Button';

describe('<Button />', () => {
  it('renders correctly', () => {
    const component = renderer.create(
      <Button>Click me</Button>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('accepts "xl" size prop', () => {
    const component = renderer.create(
      <Button size="xl">Click me</Button>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('accepts "l" size prop', () => {
    const component = renderer.create(
      <Button size="l">Click me</Button>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('accepts "s" size prop', () => {
    const component = renderer.create(
      <Button size="s">Click me</Button>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  
  it('accepts "xs" size prop', () => {
    const component = renderer.create(
      <Button size="xs">Click me</Button>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('accepts "primary" modifier prop', () => {
    const component = renderer.create(
      <Button modifier="primary">Click me</Button>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('accepts "secondary" modifier prop', () => {
    const component = renderer.create(
      <Button modifier="secondary">Click me</Button>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('accepts "anchor" modifier prop', () => {
    const component = renderer.create(
      <Button modifier="anchor">Click me</Button>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('accepts "anchor-block" modifier prop', () => {
    const component = renderer.create(
      <Button modifier="anchor-block">Click me</Button>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('accepts "disabled" prop', () => {
    const component = renderer.create(
      <Button modifier="primary" disabled>Click me</Button>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});