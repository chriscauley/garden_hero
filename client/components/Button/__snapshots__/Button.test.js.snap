// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> accepts "anchor" modifier prop 1`] = `
<button
  className="gh-btn gh-btn--anchor"
  type="button"
>
  Click me
</button>
`;

exports[`<Button /> accepts "anchor-block" modifier prop 1`] = `
<button
  className="gh-btn gh-btn--anchor-block"
  type="button"
>
  Click me
</button>
`;

exports[`<Button /> accepts "disabled" prop 1`] = `
<button
  className="gh-btn gh-btn--primary gh-btn--disabled"
  type="button"
>
  Click me
</button>
`;

exports[`<Button /> accepts "l" size prop 1`] = `
<button
  className="gh-btn gh-btn--l"
  type="button"
>
  Click me
</button>
`;

exports[`<Button /> accepts "primary" modifier prop 1`] = `
<button
  className="gh-btn gh-btn--primary"
  type="button"
>
  Click me
</button>
`;

exports[`<Button /> accepts "s" size prop 1`] = `
<button
  className="gh-btn gh-btn--s"
  type="button"
>
  Click me
</button>
`;

exports[`<Button /> accepts "secondary" modifier prop 1`] = `
<button
  className="gh-btn gh-btn--secondary"
  type="button"
>
  Click me
</button>
`;

exports[`<Button /> accepts "xl" size prop 1`] = `
<button
  className="gh-btn gh-btn--xl"
  type="button"
>
  Click me
</button>
`;

exports[`<Button /> accepts "xs" size prop 1`] = `
<button
  className="gh-btn gh-btn--xs"
  type="button"
>
  Click me
</button>
`;

exports[`<Button /> renders correctly 1`] = `
<button
  className="gh-btn"
  type="button"
>
  Click me
</button>
`;
